File: backend/app/schemas/monthly_summary.py
import uuid
from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class CategorySpendingSummary(BaseModel):
    category_id: uuid.UUID | None
    name: str
    color: str
    spent: Decimal


class BudgetProgressSummary(BaseModel):
    budget_id: uuid.UUID
    category_id: uuid.UUID
    name: str
    color: str
    limit_amount: Decimal
    spent: Decimal
    remaining: Decimal
    percent_used: Decimal | None
    overspent: bool


class MonthlySummaryResponse(BaseModel):
    month: date
    planned_income: Decimal
    actual_inflows: Decimal
    total_spending: Decimal
    available_after_spending: Decimal
    spending_percent: Decimal | None
    remaining_percent: Decimal | None
    uncategorized_count: int
    category_spending: list[CategorySpendingSummary]
    budget_progress: list[BudgetProgressSummary]


class MonthlyTrendPoint(BaseModel):
    month: date
    planned_income: Decimal
    actual_inflows: Decimal
    total_spending: Decimal


class MonthlyTrendsResponse(BaseModel):
    start_month: date
    end_month: date
    months: list[MonthlyTrendPoint]

File: frontend/src/api/monthlySummary.ts
export type CategorySpendingSummary = {
  category_id: string | null
  name: string
  color: string
  spent: string
}

export type BudgetProgressSummary = {
  budget_id: string
  category_id: string
  name: string
  color: string
  limit_amount: string
  spent: string
  remaining: string
  percent_used: string | null
  overspent: boolean
}

export type MonthlySummary = {
  month: string
  planned_income: string
  actual_inflows: string
  total_spending: string
  available_after_spending: string
  spending_percent: string | null
  remaining_percent: string | null
  uncategorized_count: number
  category_spending: CategorySpendingSummary[]
  budget_progress: BudgetProgressSummary[]
}

export async function getMonthlySummary(month: string): Promise<MonthlySummary> {
  const response = await fetch(`/api/v1/summary?month=${month}-01`)
  if (!response.ok) throw new Error('The monthly summary could not be loaded.')
  return (await response.json()) as MonthlySummary
}
